### PR TITLE
Fixed Error: local variable 'xdata' referenced before assignment" in legend_handler.py

### PR DESCRIPTION
**Pull request by kalagau**
This is the PR for #6921

### CODE REVIEW

File: lib/matplotlib/legend_handler.py [hunk 1]
```diff
@@ -148,21 +148,24 @@ def get_numpoints(self, legend):
     def get_xdata(self, legend, xdescent, ydescent, width, height, fontsize):
         numpoints = self.get_numpoints(legend)
 
-        if numpoints > 1:
+        if ((numpoints < 1) or not (isinstance(numpoints, int) or numpoints.is_integer())):
```
Comment (story645) [inline lib/matplotlib/legend_handler.py]: What's the difference between these two?
`isinstance(numpoints, int) or numpoints.is_integer())`

Comment (kalagau) [inline lib/matplotlib/legend_handler.py]: ahh they should do the same thing. Which style do you prefer?

File: lib/matplotlib/legend_handler.py [hunk 2]
```diff
@@ -148,21 +148,24 @@ def get_numpoints(self, legend):
     def get_xdata(self, legend, xdescent, ydescent, width, height, fontsize):
         numpoints = self.get_numpoints(legend)
 
-        if numpoints > 1:
+        if ((numpoints < 1) or not (numpoints.is_integer())):
```
Comment (QuLogic) [inline lib/matplotlib/legend_handler.py]: You should check for integralness first, I think. Also, most of these parentheses are unnecessary.

File: lib/matplotlib/legend_handler.py [hunk 3]
```diff
@@ -148,21 +148,24 @@ def get_numpoints(self, legend):
...
+            raise ValueError("numpoints must be a whole number and greater" +
+                 " than or equal to 1; it was", numpoints)
+        elif (numpoints == 1):
```
Comment (QuLogic) [inline lib/matplotlib/legend_handler.py]: Unnecessary parentheses.

**Comment by anntzer** [comment]
I would just have a check `if > 1: ... else: ...` (basically just remove the ==1)
Non-integer values will trigger a warning in linspace ("DeprecationWarning: object of type <class 'float'> cannot be safely interpreted as an integer.") which should be enough.
Otherwise, adding typechecks everywhere is just a never ending task.

**Comment by kalagau** [comment]
@anntzer I was just fixing it up.

**Comment by tacaswell** [comment]
@kalagau This looks like you have mixed tabs and spaces!  Can you replace them all with 4 spaces (most editors will have a tool for this (ex `untabify` in emacs) or just find and replace)?

**Comment by kalagau** [comment]
Ill add it in later today

**Comment by kalagau** [comment]
I think this should be good now! all tests are passing!

**Comment by phobson** [comment]
@kalagau Are you comfortable squashing this down to 1 commit?

**Comment by kalagau** [comment]
Sure @phobson, just did it.

**Comment by phobson** [comment]
thanks @kalagau !

**Pull request merged**